src/meta/src/barrier/context/context_impl.rs

```diff
@@ -192,14 +192,23 @@ impl CommandContext {
 
             Command::Resume => {}
 
-            Command::SourceChangeSplit(split_assignment) => {
+            Command::SourceChangeSplit(SplitState {
+                split_assignment: assignment,
+                discovered_source_splits: source_splits,
+            }) => {
                 barrier_manager_context
                     .metadata_manager
-                    .update_actor_splits_by_split_assignment(split_assignment)
+                    .update_actor_splits_by_split_assignment(assignment)
                     .await?;
+
+                barrier_manager_context
+                    .metadata_manager
+                    .update_source_splits(source_splits)
+                    .await?;
```

Comment:
I think this part should be moved before yielding the split change mutation. 

We have updated split_assignment in each related fragment in the re-assign process, which means the mutation will be applied no matter the barrier succeeds or not. 
But here we update the state table only when the barrier succeeds. It might cause inconsistency in barrier failure scenes.